test/ng/directive/ngRefSpec.js

```diff
@@ -0,0 +1,517 @@
+      var template =
+        '<div ng-if="true">' +
+          '<my-component ng-ref="myComponent"></my-component>' +
+        '</div>';
+      $compile(template)($rootScope);
+      expect($rootScope.myComponent).toBe(undefined);
+    });
+
+    it('should nullify the variable once the component is destroyed', function() {
+      var template = '<div><my-component ng-ref="myComponent"></my-component></div>';
+
+      var element = $compile(template)($rootScope);
+      expect($rootScope.myComponent).toBe(myComponentController);
+
+      var componentElement = element.children();
+      var isolateScope = componentElement.isolateScope();
+      componentElement.remove();
+      isolateScope.$destroy();
+      expect($rootScope.myComponent).toBe(null);
+    });
+
+    it('should be compatible with entering/leaving components', inject(function($animate) {
+      var template = '<my-component ng-ref="myComponent"></my-component>';
+      $rootScope.$ctrl = {};
+      var parent = $compile('<div></div>')($rootScope);
+
+      var leaving = $compile(template)($rootScope);
+      var leavingController = myComponentController;
+
+      $animate.enter(leaving, parent);
+      expect($rootScope.myComponent).toBe(leavingController);
+
+      var entering = $compile(template)($rootScope);
+      var enteringController = myComponentController;
+
+      $animate.enter(entering, parent);
+      $animate.leave(leaving, parent);
+      expect($rootScope.myComponent).toBe(enteringController);
+    }));
+
+    it('should allow binding to a nested property', function() {
+      $rootScope.obj = {};
+
+      $compile('<my-component ng-ref="obj.myComponent"></my-component>')($rootScope);
+      expect($rootScope.obj.myComponent).toBe(myComponentController);
+    });
+
+  });
+
+  it('should bind the jqlite wrapped DOM element if there is no component', inject(function($compile, $rootScope) {
+    var el = $compile('<span ng-ref="mySpan">my text</span>')($rootScope);
+
+    expect($rootScope.mySpan).toEqual(el);
+    expect($rootScope.mySpan[0].textContent).toBe('my text');
+  }));
+
+  it('should nullify the dom element value if it is destroyed', inject(function($compile, $rootScope) {
+    var element = $compile('<div><span ng-ref="mySpan">my text</span></div>')($rootScope);
+    element.children().remove();
+    expect($rootScope.mySpan).toBe(null);
+  }));
+
+  it('should bind the controller of an element directive', function() {
+    var myDirectiveController;
+
+    module(function($compileProvider) {
+      $compileProvider.directive('myDirective', function() {
+        return {
+          controller: function() {
+            myDirectiveController = this;
+          }
+        };
+      });
+    });
+
+    inject(function($compile, $rootScope) {
+      $compile('<my-directive ng-ref="myDirective"></my-directive>')($rootScope);
+
+      expect($rootScope.myDirective).toBe(myDirectiveController);
+    });
+  });
```

Comment:
Is it worth having a test that shows the element directive / component is chosen over any additional attribute directives if no `ngRefRead` attribute is available?

Reply:
will add